pkg/apb/ext_creds.go

```diff
@@ -15,77 +16,56 @@ var timeoutFreq = 6    // Seconds
 var totalTimeout = 900 // 15min
 
 // ExtractCredentials - Extract credentials from pod in a certain namespace.
-// HACK ALERT!
-// A lot of the current approach to extracting credentials and monitoring
-// output is *very* experimental and error prone. Entire approach is going
-// to be thrown out and redone asap.
 func ExtractCredentials(
 	podname string, namespace string, log *logging.Logger,
 ) (*ExtractedCredentials, error) {
-	log.Debug("Calling monitorOutput on " + podname)
-	credOut, _ := monitorOutput(podname, namespace)
-	log.Debug("oc log output: %s", string(credOut))
+	credOut := make(chan []byte)
 
-	var creds *ExtractedCredentials
-	var err error
-	creds, err = buildExtractedCredentials(credOut)
+	log.Debug("Calling monitorOutput on " + podname)
+	go monitorOutput(podname, credOut, log)
 
-	if err != nil {
-		// HACK: this is HORRIBLE. but there is definitely a time between a bind
-		// and when the container is up.
-		totalRetries := totalTimeout / timeoutFreq
-		retries := 1
-		for {
-			if retries == totalRetries {
-				errstr := "TIMED OUT WAITING FOR CONTAINER TO COME UP"
-				log.Error(errstr)
-				return nil, errors.New(errstr)
-			}
-
-			time.Sleep(time.Duration(timeoutFreq) * time.Second)
-			log.Info("Container not up yet, retrying %d of %d on pod %s", retries, totalRetries, podname)
-			credOut, _ = monitorOutput(podname, namespace)
-			log.Debug("oc log output: \n%s", string(credOut))
-			creds, err = buildExtractedCredentials(credOut)
-
-			if err == nil {
-				if creds != nil {
-					log.Debug("Pod reporting finished and returned Credentials")
-				} else {
-					log.Debug("Pod reporting finished and DID NOT return Credentials")
-				}
-				break
-			} else if err.Error() == stillWaitingError {
-				// Known error code that's received when we're either waiting for
-				// ContainerCreating, or for the pod resource to be created.
-				// These are expected states, and we'll wait until the pod is up.
-				log.Debug(err.Error())
-			} else {
-				log.Notice("WARNING: Unexpected output from apb pod")
-				log.Notice("Will keep retrying, but it's possible something has gone wrong.")
-				log.Notice(err.Error())
-			}
-
-			retries++
-		}
+	var bindOutput []byte
+	for msg := range credOut {
+		bindOutput = msg
 	}
 
-	return creds, err
+	return buildExtractedCredentials(bindOutput)
 }
 
-func monitorOutput(podname string, namespace string) ([]byte, error) {
-	return RunCommand("oc", "logs", "-f", "--namespace="+namespace, podname)
-}
+func monitorOutput(podname string, mon chan []byte, log *logging.Logger) {
+	// TODO: Error handling here
+	// It would also be nice to gather the script output that exec runs
+	// instead of only getting the credentials
+	retries := 20
+
+	for r := 1; r <= retries; r++ {
+		output, _ := RunCommand("oc", "exec", podname, "broker-bind-creds")
+
```

Comment:
output will contain the error. I use it on line 45.  Getting the other param will get ```exit status 1``` from the command. I think it's because we use CombinedOutput ```output, err := exec.Command(cmd, args...).CombinedOutput()```